File: data_baseline/openrouter__qwen__qwen3.7-plus/scala/rep7/Main.scala
//> using scala 3.3.1
//> using dep org.http4s::http4s-ember-server:0.23.34
//> using dep org.http4s::http4s-dsl:0.23.34
//> using dep org.http4s::http4s-circe:0.23.34
//> using dep io.circe::circe-core:0.14.15

import org.http4s._
import org.http4s.dsl.io._
import org.http4s.implicits._
import org.http4s.circe._
import org.http4s.ember.server.EmberServerBuilder
import org.http4s.headers.`WWW-Authenticate`
import org.http4s.Challenge
import io.circe._
import io.circe.syntax._
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.UUID
import scala.collection.concurrent.TrieMap
import cats.effect._
import com.comcast.ip4s._

case class User(id: Int, username: String, password: String)
case class Todo(
  id: Int,
  userId: Int,
  title: String,
  description: String,
  completed: Boolean,
  createdAt: String,
  updatedAt: String
)

case class UserResponse(id: Int, username: String)
object UserResponse {
  given Encoder[UserResponse] = Encoder.instance { u =>
    Json.obj(
      "id" -> Json.fromInt(u.id),
      "username" -> Json.fromString(u.username)
    )
  }
}

case class TodoResponse(
  id: Int,
  title: String,
  description: String,
  completed: Boolean,
  created_at: String,
  updated_at: String
)
object TodoResponse {
  given Encoder[TodoResponse] = Encoder.instance { t =>
    Json.obj(
      "id" -> Json.fromInt(t.id),
      "title" -> Json.fromString(t.title),
      "description" -> Json.fromString(t.description),
      "completed" -> Json.fromBoolean(t.completed),
      "created_at" -> Json.fromString(t.created_at),
      "updated_at" -> Json.fromString(t.updated_at)
    )
  }
}

case class ErrorResponse(error: String)
object ErrorResponse {
  given Encoder[ErrorResponse] = Encoder.instance { e =>
    Json.obj("error" -> Json.fromString(e.error))
  }
}

case class RegisterRequest(username: Option[String], password: Option[String])
object RegisterRequest {
  given Decoder[RegisterRequest] = Decoder.instance { c =>
    for {
      username <- c.downField("username").as[Option[String]]
      password <- c.downField("password").as[Option[String]]
    } yield RegisterRequest(username, password)
  }
  given EntityDecoder[IO, RegisterRequest] = jsonOf[IO, RegisterRequest]
}

case class LoginRequest(username: Option[String], password: Option[String])
object LoginRequest {
  given Decoder[LoginRequest] = Decoder.instance { c =>
    for {
      username <- c.downField("username").as[Option[String]]
      password <- c.downField("password").as[Option[String]]
    } yield LoginRequest(username, password)
  }
  given EntityDecoder[IO, LoginRequest] = jsonOf[IO, LoginRequest]
}

case class PasswordChangeRequest(oldPassword: Option[String], newPassword: Option[String])
object PasswordChangeRequest {
  given Decoder[PasswordChangeRequest] = Decoder.instance { c =>
    for {
      oldPassword <- c.downField("old_password").as[Option[String]]
      newPassword <- c.downField("new_password").as[Option[String]]
    } yield PasswordChangeRequest(oldPassword, newPassword)
  }
  given EntityDecoder[IO, PasswordChangeRequest] = jsonOf[IO, PasswordChangeRequest]
}

case class CreateTodoRequest(title: Option[String], description: Option[String])
object CreateTodoRequest {
  given Decoder[CreateTodoRequest] = Decoder.instance { c =>
    for {
      title <- c.downField("title").as[Option[String]]
      desc <- c.downField("description").as[Option[String]]
    } yield CreateTodoRequest(title, desc)
  }
  given EntityDecoder[IO, CreateTodoRequest] = jsonOf[IO, CreateTodoRequest]
}

case class UpdateTodoRequest(title: Option[String], description: Option[String], completed: Option[Boolean])
object UpdateTodoRequest {
  given Decoder[UpdateTodoRequest] = Decoder.instance { c =>
    for {
      title <- c.downField("title").as[Option[String]]
      desc <- c.downField("description").as[Option[String]]
      completed <- c.downField("completed").as[Option[Boolean]]
    } yield UpdateTodoRequest(title, desc, completed)
  }
  given EntityDecoder[IO, UpdateTodoRequest] = jsonOf[IO, UpdateTodoRequest]
}

object State {
  val users = TrieMap[Int, User]()
  val sessions = TrieMap[String, Int]()
  val todos = TrieMap[Int, Todo]()
  
  @volatile var userIdCounter = 0
  @volatile var todoIdCounter = 0
  
  def nextUserId(): Int = {
    userIdCounter += 1
    userIdCounter
  }
  
  def nextTodoId(): Int = {
    todoIdCounter += 1
    todoIdCounter
  }
}

def now(): String = {
  val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
  ZonedDateTime.now(ZoneOffset.UTC).format(formatter)
}

object Main extends IOApp {
  def getSessionId(req: Request[IO]): Option[String] = {
    req.cookies.find(_.name == "session_id").map(_.content)
  }

  def requireAuth(req: Request[IO]): IO[Option[User]] = {
    getSessionId(req) match {
      case Some(token) =>
        IO(State.sessions.get(token).flatMap(userId => State.users.get(userId)))
      case None => IO(None)
    }
  }

  val authChallenge = `WWW-Authenticate`(Challenge("Bearer", "todo-app"))

  def unauthResp: IO[Response[IO]] = Unauthorized(authChallenge, ErrorResponse("Authentication required").asJson)
  def invalidCredsResp: IO[Response[IO]] = Unauthorized(authChallenge, ErrorResponse("Invalid credentials").asJson)

  val httpApp = HttpRoutes.of[IO] {
    case req @ POST -> Root / "register" =>
      for {
        reqBody <- req.as[RegisterRequest]
        resp <- {
          val uName = reqBody.username.getOrElse("")
          val pWord = reqBody.password.getOrElse("")
          if (!uName.matches("^[a-zA-Z0-9_]+$") || uName.length < 3 || uName.length > 50) {
            BadRequest(ErrorResponse("Invalid username").asJson)
          } else if (pWord.length < 8) {
            BadRequest(ErrorResponse("Password too short").asJson)
          } else {
            if (State.users.values.exists(_.username == uName)) {
              Conflict(ErrorResponse("Username already exists").asJson)
            } else {
              val id = State.nextUserId()
              val newUser = User(id, uName, pWord)
              State.users.put(id, newUser)
              Created(UserResponse(id, newUser.username).asJson)
            }
          }
        }
      } yield resp

    case req @ POST -> Root / "login" =>
      for {
        reqBody <- req.as[LoginRequest]
        resp <- {
          val uName = reqBody.username.getOrElse("")
          val pWord = reqBody.password.getOrElse("")
          State.users.values.find(u => u.username == uName && u.password == pWord) match {
            case Some(user) =>
              val token = UUID.randomUUID().toString
              State.sessions.put(token, user.id)
              val cookie = ResponseCookie("session_id", token, httpOnly = true, path = Some("/"))
              Ok(UserResponse(user.id, user.username).asJson).map(_.addCookie(cookie))
            case None =>
              invalidCredsResp
          }
        }
      } yield resp

    case req @ POST -> Root / "logout" =>
      for {
        user <- requireAuth(req)
        resp <- user match {
          case Some(_) =>
            getSessionId(req) match {
              case Some(token) =>
                State.sessions.remove(token)
                val clearCookie = ResponseCookie("session_id", "", maxAge = Some(0L), path = Some("/"))
                Ok(Json.obj()).map(_.addCookie(clearCookie))
              case None =>
                Ok(Json.obj())
            }
          case None =>
            unauthResp
        }
      } yield resp

    case req @ GET -> Root / "me" =>
      for {
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) => Ok(UserResponse(u.id, u.username).asJson)
          case None => unauthResp
        }
      } yield resp

    case req @ PUT -> Root / "password" =>
      for {
        reqBody <- req.as[PasswordChangeRequest]
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) =>
            val oldP = reqBody.oldPassword.getOrElse("")
            val newP = reqBody.newPassword.getOrElse("")
            if (u.password != oldP) {
              invalidCredsResp
            } else if (newP.length < 8) {
              BadRequest(ErrorResponse("Password too short").asJson)
            } else {
              val updatedUser = u.copy(password = newP)
              State.users.put(u.id, updatedUser)
              Ok(Json.obj())
            }
          case None =>
            unauthResp
        }
      } yield resp

    case req @ GET -> Root / "todos" =>
      for {
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) =>
            val userTodos = State.todos.values
              .filter(_.userId == u.id)
              .toList
              .sortBy(_.id)
              .map(t => TodoResponse(t.id, t.title, t.description, t.completed, t.createdAt, t.updatedAt))
            Ok(userTodos.asJson)
          case None => unauthResp
        }
      } yield resp

    case req @ POST -> Root / "todos" =>
      for {
        reqBody <- req.as[CreateTodoRequest]
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) =>
            val title = reqBody.title.getOrElse("")
            val desc = reqBody.description.getOrElse("")
            if (title.isEmpty) {
              BadRequest(ErrorResponse("Title is required").asJson)
            } else {
              val id = State.nextTodoId()
              val ts = now()
              val newTodo = Todo(
                id = id,
                userId = u.id,
                title = title,
                description = desc,
                completed = false,
                createdAt = ts,
                updatedAt = ts
              )
              State.todos.put(id, newTodo)
              Created(TodoResponse(newTodo.id, newTodo.title, newTodo.description, newTodo.completed, newTodo.createdAt, newTodo.updatedAt).asJson)
            }
          case None => unauthResp
        }
      } yield resp

    case req @ GET -> Root / "todos" / IntVar(todoId) =>
      for {
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) =>
            State.todos.get(todoId) match {
              case Some(t) if t.userId == u.id =>
                Ok(TodoResponse(t.id, t.title, t.description, t.completed, t.createdAt, t.updatedAt).asJson)
              case _ =>
                NotFound(ErrorResponse("Todo not found").asJson)
            }
          case None => unauthResp
        }
      } yield resp

    case req @ PUT -> Root / "todos" / IntVar(todoId) =>
      for {
        reqBody <- req.as[UpdateTodoRequest]
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) =>
            State.todos.get(todoId) match {
              case Some(t) if t.userId == u.id =>
                if (reqBody.title.exists(_.isEmpty)) {
                  BadRequest(ErrorResponse("Title is required").asJson)
                } else {
                  val newTitle = reqBody.title.getOrElse(t.title)
                  val newDesc = reqBody.description.getOrElse(t.description)
                  val newCompleted = reqBody.completed.getOrElse(t.completed)
                  val updatedTodo = t.copy(
                    title = newTitle,
                    description = newDesc,
                    completed = newCompleted,
                    updatedAt = now()
                  )
                  State.todos.put(todoId, updatedTodo)
                  Ok(TodoResponse(updatedTodo.id, updatedTodo.title, updatedTodo.description, updatedTodo.completed, updatedTodo.createdAt, updatedTodo.updatedAt).asJson)
                }
              case _ =>
                NotFound(ErrorResponse("Todo not found").asJson)
            }
          case None => unauthResp
        }
      } yield resp

    case req @ DELETE -> Root / "todos" / IntVar(todoId) =>
      for {
        user <- requireAuth(req)
        resp <- user match {
          case Some(u) =>
            State.todos.get(todoId) match {
              case Some(t) if t.userId == u.id =>
                State.todos.remove(todoId)
                NoContent()
              case _ =>
                NotFound(ErrorResponse("Todo not found").asJson)
            }
          case None => unauthResp
        }
      } yield resp
  }.orNotFound

  def run(args: List[String]): IO[ExitCode] = {
    val portInt = args match {
      case "--port" :: p :: _ => p.toInt
      case _ => 8080
    }
    val p = Port.fromInt(portInt).getOrElse(port"8080")
    
    EmberServerBuilder.default[IO]
      .withHost(ipv4"0.0.0.0")
      .withPort(p)
      .withHttpApp(httpApp)
      .build
      .use(_ => IO.never)
      .as(ExitCode.Success)
  }
}

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep7/src/main.rs
use axum::{
    extract::{Json, Path, State},
    http::{HeaderMap, StatusCode},
    response::IntoResponse,
    routing::{delete, get, post, put},
    Router,
};
use bcrypt::{hash, verify, DEFAULT_COST};
use chrono::Utc;
use clap::Parser;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use uuid::Uuid;

#[derive(Parser)]
#[command(name = "todo_app")]
#[command(about = "A REST API server for managing personal todo items")]
struct Cli {
    #[arg(long, default_value_t = 8080)]
    port: u16,
}

#[derive(Clone)]
struct AppState {
    users: Arc<Mutex<HashMap<u64, User>>>,
    todos: Arc<Mutex<HashMap<u64, Todo>>>,
    sessions: Arc<Mutex<HashMap<String, u64>>>,
    username_to_id: Arc<Mutex<HashMap<String, u64>>>,
    next_user_id: Arc<Mutex<u64>>,
    next_todo_id: Arc<Mutex<u64>>,
}

#[derive(Clone, Serialize, Deserialize)]
struct User {
    id: u64,
    username: String,
    password_hash: String,
}

#[derive(Clone, Serialize, Deserialize)]
struct Todo {
    id: u64,
    user_id: u64,
    title: String,
    description: String,
    completed: bool,
    created_at: String,
    updated_at: String,
}

impl AppState {
    fn new() -> Self {
        Self {
            users: Arc::new(Mutex::new(HashMap::new())),
            todos: Arc::new(Mutex::new(HashMap::new())),
            sessions: Arc::new(Mutex::new(HashMap::new())),
            username_to_id: Arc::new(Mutex::new(HashMap::new())),
            next_user_id: Arc::new(Mutex::new(1)),
            next_todo_id: Arc::new(Mutex::new(1)),
        }
    }
}

fn get_session_id(headers: &HeaderMap) -> Option<String> {
    let cookie_header = headers.get(axum::http::header::COOKIE)?.to_str().ok()?;
    for cookie in cookie_header.split(';') {
        let cookie = cookie.trim();
        if let Some((name, value)) = cookie.split_once('=') {
            if name == "session_id" {
                return Some(value.to_string());
            }
        }
    }
    None
}

fn is_valid_username(username: &str) -> bool {
    if username.len() < 3 || username.len() > 50 {
        return false;
    }
    username.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

type AppResult<T> = Result<T, (StatusCode, Json<serde_json::Value>)>;

#[derive(Deserialize)]
struct RegisterRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn register(
    State(state): State<AppState>,
    Json(req): Json<RegisterRequest>,
) -> AppResult<impl IntoResponse> {
    let username = req.username.ok_or_else(|| (StatusCode::BAD_REQUEST, Json(json!({"error": "Invalid username"}))))?;
    let password = req.password.ok_or_else(|| (StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))))?;

    if !is_valid_username(&username) {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Invalid username"}))));
    }
    if password.len() < 8 {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))));
    }

    let mut username_to_id = state.username_to_id.lock().unwrap();
    if username_to_id.contains_key(&username) {
        return Err((StatusCode::CONFLICT, Json(json!({"error": "Username already exists"}))));
    }

    let mut users = state.users.lock().unwrap();
    let mut next_user_id = state.next_user_id.lock().unwrap();
    
    let id = *next_user_id;
    *next_user_id += 1;
    
    let password_hash = hash(&password, DEFAULT_COST)
        .map_err(|_| (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({"error": "Internal server error"}))))?;

    let user = User {
        id,
        username: username.clone(),
        password_hash,
    };

    users.insert(id, user);
    username_to_id.insert(username.clone(), id);

    Ok((StatusCode::CREATED, Json(json!({"id": id, "username": username}))))
}

#[derive(Deserialize)]
struct LoginRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn login(
    State(state): State<AppState>,
    Json(req): Json<LoginRequest>,
) -> AppResult<(StatusCode, HeaderMap, Json<serde_json::Value>)> {
    let username = req.username.unwrap_or_default();
    let password = req.password.unwrap_or_default();

    let username_to_id = state.username_to_id.lock().unwrap();
    let user_id = *username_to_id.get(&username).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))))?;
    
    let users = state.users.lock().unwrap();
    let user = users.get(&user_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))))?;

    let valid = verify(&password, &user.password_hash)
        .map_err(|_| (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({"error": "Internal server error"}))))?;
    
    if !valid {
        return Err((StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))));
    }

    let session_id = Uuid::new_v4().to_string();
    let mut sessions = state.sessions.lock().unwrap();
    sessions.insert(session_id.clone(), user_id);

    let cookie = format!("session_id={}; Path=/; HttpOnly", session_id);
    let mut headers = HeaderMap::new();
    headers.insert(
        axum::http::header::SET_COOKIE,
        axum::http::HeaderValue::from_str(&cookie).unwrap(),
    );

    Ok((StatusCode::OK, headers, Json(json!({"id": user.id, "username": user.username}))))
}

async fn logout(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let mut sessions = state.sessions.lock().unwrap();
    sessions.remove(&session_id);

    Ok((StatusCode::OK, Json(json!({}))))
}

async fn get_me(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let users = state.users.lock().unwrap();
    let user = users.get(&user_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;

    Ok((StatusCode::OK, Json(json!({"id": user.id, "username": user.username}))))
}

#[derive(Deserialize)]
struct UpdatePasswordRequest {
    old_password: Option<String>,
    new_password: Option<String>,
}

async fn update_password(
    headers: HeaderMap,
    State(state): State<AppState>,
    Json(req): Json<UpdatePasswordRequest>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;

    let old_password = req.old_password.unwrap_or_default();
    let new_password = req.new_password.unwrap_or_default();

    if new_password.len() < 8 {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))));
    }

    let mut users = state.users.lock().unwrap();
    let user = users.get_mut(&user_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))))?;

    let valid = verify(&old_password, &user.password_hash)
        .map_err(|_| (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({"error": "Internal server error"}))))?;
    
    if !valid {
        return Err((StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))));
    }

    let new_hash = hash(&new_password, DEFAULT_COST)
        .map_err(|_| (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({"error": "Internal server error"}))))?;
    
    user.password_hash = new_hash;

    Ok((StatusCode::OK, Json(json!({}))))
}

async fn get_todos(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let todos = state.todos.lock().unwrap();
    let mut user_todos: Vec<_> = todos.values()
        .filter(|t| t.user_id == user_id)
        .cloned()
        .collect();
    
    user_todos.sort_by_key(|t| t.id);
    
    let response_todos: Vec<serde_json::Value> = user_todos.into_iter().map(|t| json!({
        "id": t.id,
        "title": t.title,
        "description": t.description,
        "completed": t.completed,
        "created_at": t.created_at,
        "updated_at": t.updated_at,
    })).collect();
    
    Ok((StatusCode::OK, Json(serde_json::Value::Array(response_todos))))
}

#[derive(Deserialize)]
struct CreateTodoRequest {
    title: Option<String>,
    description: Option<String>,
}

async fn create_todo(
    headers: HeaderMap,
    State(state): State<AppState>,
    Json(req): Json<CreateTodoRequest>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;

    let title = req.title.ok_or_else(|| (StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))))?;
    if title.trim().is_empty() {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))));
    }
    
    let description = req.description.unwrap_or_default();

    let mut todos = state.todos.lock().unwrap();
    let mut next_todo_id = state.next_todo_id.lock().unwrap();
    
    let id = *next_todo_id;
    *next_todo_id += 1;
    
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    
    let todo = Todo {
        id,
        user_id,
        title: title.clone(),
        description: description.clone(),
        completed: false,
        created_at: now.clone(),
        updated_at: now.clone(),
    };
    
    todos.insert(id, todo.clone());
    
    Ok((StatusCode::CREATED, Json(json!({
        "id": todo.id,
        "title": todo.title,
        "description": todo.description,
        "completed": todo.completed,
        "created_at": todo.created_at,
        "updated_at": todo.updated_at,
    }))))
}

async fn get_todo(
    headers: HeaderMap,
    Path(todo_id): Path<u64>,
    State(state): State<AppState>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;

    let todos = state.todos.lock().unwrap();
    let todo = todos.get(&todo_id).ok_or_else(|| (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))))?;
    
    if todo.user_id != user_id {
        return Err((StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))));
    }
    
    Ok((StatusCode::OK, Json(json!({
        "id": todo.id,
        "title": todo.title,
        "description": todo.description,
        "completed": todo.completed,
        "created_at": todo.created_at,
        "updated_at": todo.updated_at,
    }))))
}

#[derive(Deserialize)]
struct UpdateTodoRequest {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

async fn update_todo(
    headers: HeaderMap,
    Path(todo_id): Path<u64>,
    State(state): State<AppState>,
    Json(req): Json<UpdateTodoRequest>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;

    let mut todos = state.todos.lock().unwrap();
    let todo = todos.get_mut(&todo_id).ok_or_else(|| (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))))?;
    
    if todo.user_id != user_id {
        return Err((StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))));
    }

    if let Some(ref title) = req.title {
        if title.trim().is_empty() {
            return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))));
        }
        todo.title = title.clone();
    }
    
    if let Some(ref description) = req.description {
        todo.description = description.clone();
    }
    
    if let Some(completed) = req.completed {
        todo.completed = completed;
    }
    
    todo.updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    
    Ok((StatusCode::OK, Json(json!({
        "id": todo.id,
        "title": todo.title,
        "description": todo.description,
        "completed": todo.completed,
        "created_at": todo.created_at,
        "updated_at": todo.updated_at,
    }))))
}

async fn delete_todo(
    headers: HeaderMap,
    Path(todo_id): Path<u64>,
    State(state): State<AppState>,
) -> AppResult<impl IntoResponse> {
    let session_id = get_session_id(&headers).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;
    
    let sessions = state.sessions.lock().unwrap();
    let user_id = *sessions.get(&session_id).ok_or_else(|| (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))))?;

    let mut todos = state.todos.lock().unwrap();
    let todo = todos.get(&todo_id).ok_or_else(|| (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))))?;
    
    if todo.user_id != user_id {
        return Err((StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))));
    }

    todos.remove(&todo_id);
    
    Ok(StatusCode::NO_CONTENT)
}

#[tokio::main]
async fn main() {
    let cli = Cli::parse();
    let state = AppState::new();
    
    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(get_me))
        .route("/password", put(update_password))
        .route("/todos", get(get_todos))
        .route("/todos", post(create_todo))
        .route("/todos/:id", get(get_todo))
        .route("/todos/:id", put(update_todo))
        .route("/todos/:id", delete(delete_todo))
        .with_state(state);

    let addr = format!("0.0.0.0:{}", cli.port);
    let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
    println!("Listening on {}", addr);
    axum::serve(listener, app).await.unwrap();
}